proposals/src/header.rs

```diff
@@ -0,0 +1,573 @@
+    #[must_use]
+    pub const fn to_bytes(self) -> [u8; Self::LENGTH] {
+        self.to_u16().to_be_bytes()
+    }
+    /// Get the underlying bytes of `ChainType`.
+    #[must_use]
+    pub const fn into_bytes(self) -> [u8; Self::LENGTH] {
+        self.to_bytes()
+    }
+}
+
+impl From<ChainType> for [u8; ChainType::LENGTH] {
+    fn from(chain_type: ChainType) -> Self {
+        chain_type.to_bytes()
+    }
+}
+
+impl From<[u8; ChainType::LENGTH]> for ChainType {
+    fn from(bytes: [u8; ChainType::LENGTH]) -> Self {
+        let v = u16::from_be_bytes(bytes);
+        match v {
+            0x0100 => ChainType::Evm,
+            0x0200 => ChainType::Substrate,
+            0x0301 => ChainType::PolkadotRelayChain,
+            0x0302 => ChainType::KusamaRelayChain,
+            0x0400 => ChainType::Cosmos,
+            0x0500 => ChainType::Solana,
+            _ => ChainType::Unknown,
+        }
+    }
+}
+
+impl ChainId {
+    /// Length of the `ChainId` in bytes.
+    pub const LENGTH: usize = core::mem::size_of::<u32>();
+
+    /// Convert `ChainId` to `u32`.
+    #[must_use]
+    pub const fn to_u32(self) -> u32 {
+        self.0
+    }
+
+    /// Get the underlying bytes of `ChainId`.
+    #[must_use]
+    pub const fn to_bytes(self) -> [u8; Self::LENGTH] {
+        self.0.to_be_bytes()
+    }
+    /// Get the underlying bytes of `ChainId`.
+    #[must_use]
+    pub const fn into_bytes(self) -> [u8; Self::LENGTH] {
+        self.to_bytes()
+    }
+}
+
+impl From<ChainId> for [u8; ChainId::LENGTH] {
+    fn from(chain_id: ChainId) -> Self {
+        chain_id.to_bytes()
+    }
+}
+
+impl From<[u8; ChainId::LENGTH]> for ChainId {
+    fn from(bytes: [u8; ChainId::LENGTH]) -> Self {
+        let v = u32::from_be_bytes(bytes);
+        ChainId(v)
+    }
+}
+
+impl From<u32> for ChainId {
+    fn from(v: u32) -> Self {
+        ChainId(v)
+    }
+}
+
+impl From<ChainId> for u32 {
+    fn from(chain_id: ChainId) -> Self {
+        chain_id.0
+    }
+}
+
+impl ResourceId {
+    /// Length of the `ResourceId` (32 bytes).
+    pub const LENGTH: usize = 32;
+    /// Create a new `ResourceId`.
+    #[must_use]
+    pub fn new(
+        target_system: TargetSystem,
+        chain_type: ChainType,
+        chain_id: ChainId,
+    ) -> Self {
+        let mut bytes = [0u8; Self::LENGTH];
+        let target_system_bytes: [u8; TargetSystem::LENGTH] =
+            target_system.into();
+        let f = 0;
+        let t = TargetSystem::LENGTH;
+        bytes[f..t].copy_from_slice(&target_system_bytes);
+        let f = t;
+        let t = t + ChainType::LENGTH;
+        bytes[f..t].copy_from_slice(&chain_type.into_bytes());
+        let f = t;
+        let t = t + ChainId::LENGTH;
+        bytes[f..t].copy_from_slice(&chain_id.into_bytes());
+        Self(bytes)
+    }
+
+    /// Get the [`TargetSystem`] from the `ResourceId`.
+    ///
+    /// The `TargetSystem` is the first 26 bytes of the `ResourceId`.
```

Comment:
@saiakilesh please confirm here as well. @shekohex I believe the `TargetSystem` is the LAST 26 bytes. Please try not to change things and keep it consistent with what is currently in `dkg-test-suite`

Reply:
https://github.com/webb-tools/dkg-substrate/blob/master/dkg-test-suite/src/evm/util/utils.ts#L517-L532

Reply:
> My comment from our conversation for future readers.

Yes, you are correct, but this is not always the case, the first 26 bytes would be always the target system, it could be a contract address left-padded with zeros (6 zeros + contract address 20 bytes) *or* it could be a tree id, and in this case, it will be left-padded with 22 zeros! (22 zeros + tree id which is 4 bytes).
if you have a better idea to rephrase this in a simpler way that would be good! as we need it to be clear as possible in this implementation.